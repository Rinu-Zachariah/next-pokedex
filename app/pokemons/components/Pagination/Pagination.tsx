'use client';
import React, { useCallback } from 'react';
import { useRouter } from 'next/navigation';
import { ChevronLeft, ChevronRight, ChevronsRight, ChevronsLeft } from "lucide-react";

interface IProps {
  total: number;
  page: number;
  pageSize: number;
  filterName: string;
}

const Pagination: React.FC<IProps> = ({
  total, page, pageSize, filterName
}) => {
  const router = useRouter();
  const lastPage = Math.ceil(total / pageSize);

  const goToPage = useCallback((newPage: number) => {
    const params = new URLSearchParams();
    if (filterName) {
      params.set('name', filterName);
    }
    params.set('page', String(newPage));
    router.replace(`?${params.toString()}`);
  }, [router, filterName]);

  if (lastPage <= 1) {
    return null;
  }

  return (
    <div className="flex justify-center items-center gap-2 mb-6">
      <button
        onClick={() => goToPage(1)}
        disabled={page === 1}
        className={
          `p-2 rounded-lg ${page === 1 ? 
            'bg-gray-300 text-gray-400 cursor-not-allowed' : 
            'bg-blue-900 text-white hover:bg-blue-800'
          }`
        }
      >
        <ChevronsLeft size={20} />
      </button>
      <button
        onClick={() => goToPage(page - 1)}
        disabled={page <= 1}
        className={
          `p-2 rounded-lg ${page <= 1 ? 
            'bg-gray-300 text-gray-400 cursor-not-allowed' :
            'bg-blue-900 text-white hover:bg-blue-800'
          }`
        }
      >
        <ChevronLeft size={20} />
      </button>
      <span>
        Page <b>{page}</b> of {lastPage}
      </span>
      <button
        onClick={() => goToPage(page + 1)}
        disabled={page === lastPage}
        className={
          `p-2 rounded-lg ${page === lastPage ?
            'bg-gray-300 text-gray-400 cursor-not-allowed' :
            'bg-blue-900 text-white hover:bg-blue-800'
          }`
        }
      >
        <ChevronRight size={20} />
      </button>
      <button
        onClick={() => goToPage(lastPage)}
        disabled={page === lastPage}
        className={
          `p-2 rounded-lg ${page === lastPage ?
            'bg-gray-300 text-gray-400 cursor-not-allowed' :
            'bg-blue-900 text-white hover:bg-blue-800'
          }`
        }
      >
        <ChevronsRight size={20} />
      </button>
    </div>
  );
};

export default Pagination;