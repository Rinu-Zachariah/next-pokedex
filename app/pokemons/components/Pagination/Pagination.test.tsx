import React from 'react';
import { render } from '@testing-library/react';
import Pagination from './Pagination';

jest.mock('next/navigation', () => ({
  useRouter: () => ({
    push: jest.fn(),
    replace: jest.fn(),
    prefetch: jest.fn(),
    refresh: jest.fn(),
    back: jest.fn(),
    forward: jest.fn(),
  }),
  usePathname: () => '/',
  useSearchParams: () => ({
    get: jest.fn(),
    set: jest.fn(),
  }),
}));

describe('Pagination', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<Pagination total={60} page={2} pageSize={20} filterName="charizard" />);
    expect(asFragment()).toMatchSnapshot();
  });
});
