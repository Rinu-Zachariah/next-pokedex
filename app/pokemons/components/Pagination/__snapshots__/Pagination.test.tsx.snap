// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly 1`] = `
<DocumentFragment>
  <div
    class="flex justify-center items-center gap-2 mb-6"
  >
    <button
      class="p-2 rounded-lg bg-blue-900 text-white hover:bg-blue-800"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-chevrons-left"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m11 17-5-5 5-5"
        />
        <path
          d="m18 17-5-5 5-5"
        />
      </svg>
    </button>
    <button
      class="p-2 rounded-lg bg-blue-900 text-white hover:bg-blue-800"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-chevron-left"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m15 18-6-6 6-6"
        />
      </svg>
    </button>
    <span>
      Page 
      <b>
        2
      </b>
       of 3
    </span>
    <button
      class="p-2 rounded-lg bg-blue-900 text-white hover:bg-blue-800"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-chevron-right"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m9 18 6-6-6-6"
        />
      </svg>
    </button>
    <button
      class="p-2 rounded-lg bg-blue-900 text-white hover:bg-blue-800"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-chevrons-right"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m6 17 5-5-5-5"
        />
        <path
          d="m13 17 5-5-5-5"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
